### PR TITLE
Mean computation in linalg - VectorMean

### PR DESCRIPTION
**Pull request by OXPHOS**
Issue #3051 
- Added VectorMean and Mean implementation in Redux Module. Modified from and implemented VectorSum.h and Sum.h
- Can return means for whole matrix, symmetric matrix, matrix blocks and return columnwise and rowwise means. 
- Diag values can be excluded for mean calculation in Eigen3 and in square matrices in Viennacl. 
- Empty matrices/vectors considered.
- Compile test passed

- Corrected typos in Redux.h (comments) 

I should have run tests before pulling. However, I find it a little bit difficult for me. I will be appreciated if someone can tell me how to run tests. Thanks!

### CODE REVIEW

**Comment by OXPHOS** [comment]
I might need some hints about the errors if possible. Thanks in advance!

**Comment by karlnapf** [comment]
run cmake with `-DENABLE_TESTING=ON` and then run `make test` or just `make unit-tests`

File: src/shogun/mathematics/linalg/internal/implementation/Mean.h [hunk 1]
```diff
@@ -0,0 +1,984 @@
+/*
+ * Copyright (c) The Shogun Machine Learning Toolbox
+ * Written (w) 2014 Soumyajit De
+ * Written (w) 2014 Khaled Nasr
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/implementation/Mean.h]: you can add your name here :)

File: src/shogun/mathematics/linalg/internal/implementation/Mean.h [hunk 2]
```diff
@@ -0,0 +1,984 @@
...
+#include <shogun/mathematics/linalg/internal/implementation/Sum.h>
+
+#include <shogun/mathematics/eigen3.h>
+
+#ifdef HAVE_VIENNACL
+#include <shogun/mathematics/linalg/internal/implementation/Scale.h>
+#include <shogun/mathematics/linalg/internal/opencl_util.h>
+#include <shogun/lib/GPUMatrix.h>
+#include <shogun/lib/GPUVector.h>
+
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/implementation/Mean.h]: no newline

File: src/shogun/mathematics/linalg/internal/implementation/Mean.h [hunk 3]
```diff
@@ -0,0 +1,984 @@
...
+{
+	/** Scalar type */
+	typedef typename Matrix::Scalar T;
+
+	/**
+	 * Method that computes the mean of a symmetric dense matrix
+	 *
+	 * @param m the matrix whose mean has to be computed
+	 * @param no_diag if true, diagonal entries are excluded from the mean
+	 * @return the mean computed as \f$\mean_{i,j}m_{i,j}\f$
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/implementation/Mean.h]: I think latex doesnt have \mean....use \frac{1}{N^2}\sum_{i,j}, or \frac{1}{N(N-1)}\sum_{i,j} if diagonal is avoided

**Comment by karlnapf** [comment]
I have to run now ... will finish reviewing this later. Thanks for the patch :)
@lambday can you also have a look?

File: src/shogun/mathematics/linalg/internal/implementation/Mean.h (hunk 3 above)
Comment (OXPHOS) [inline src/shogun/mathematics/linalg/internal/implementation/Mean.h]: Oops..Sorry I had no experience with Latex and I just did a simple replacement.
Will work on the code. Thanks!

File: src/shogun/mathematics/linalg/internal/implementation/Mean.h [hunk 4]
```diff
@@ -0,0 +1,984 @@
...
+	/**
+	 * Method that computes the mean of SGMatrix using Eigen3
+	 *
+	 * @param m the matrix whose mean has to be computed
+	 * @param no_diag if true, diagonal entries are excluded from the mean
+	 * @return the mean computed as \f$\mean_{i,j}m_{i,j}\f$
+	 */
+	static T compute(SGMatrix<T> mat, bool no_diag)
+	{
+                REQUIRE(mat.num_rows > 0, "Matrix can not be empty!\n");
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/implementation/Mean.h]: I guess you should also check the columns here?

File: src/shogun/mathematics/linalg/internal/implementation/Mean.h [hunk 5]
```diff
@@ -0,0 +1,984 @@
...
+{
+	/** Scalar type */
+	typedef typename Matrix::Scalar T;
+
+	/**
+	 * Method that computes the mean of SGMatrix using Eigen3
+	 *
+	 * @param m the matrix whose mean has to be computed
+	 * @param no_diag if true, diagonal entries are excluded from the mean
+	 * @return the mean computed as \f$\mean_{i,j}m_{i,j}\f$
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/implementation/Mean.h]: latex

File: src/shogun/mathematics/linalg/internal/implementation/Mean.h [hunk 6]
```diff
@@ -0,0 +1,984 @@
...
+	 */
+	static T compute(SGMatrix<T> mat, bool no_diag)
+	{
+                REQUIRE(mat.num_rows > 0, "Matrix can not be empty!\n");
+                
+                // remove the main diagonal elements if required
+                if (no_diag) 
+                {
+                    index_t len_major_diag = mat.rows() < mat.cols() ? mat.rows() : mat.cols();
+                    return (sum<Backend::EIGEN3, SGMatrix<T> >::compute(mat, no_diag)
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/implementation/Mean.h]: I like it that you re-use the sum module here :)

File: src/shogun/mathematics/linalg/internal/implementation/Mean.h [hunk 7]
```diff
@@ -0,0 +1,984 @@
...
+                }
+	}
+};
+
+/**
+ * @brief Specialization of generic colwise_mean which works with SGMatrix and uses Eigen3
+ * as backend for computing mean.
+ */
+template <class Matrix>
+struct colwise_mean<Backend::EIGEN3,Matrix> ////////////////// SGVector rowwise divide
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/implementation/Mean.h]: we dont do such type of comments usually

File: src/shogun/mathematics/linalg/internal/implementation/Mean.h (hunk 4 above)
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/implementation/Mean.h]: sorry it is row-wise I did not realise

File: src/shogun/mathematics/linalg/internal/implementation/Mean.h [hunk 8]
```diff
@@ -0,0 +1,984 @@
...
+	/**
+	 * Method that computes the column wise mean of SGMatrix using Eigen3
+	 *
+	 * @param m the matrix whose colwise mean has to be computed
+	 * @param no_diag if true, diagonal entries are excluded from the mean
+	 * @return the colwise mean computed as \f$s_j=\mean_{i}m_{i,j}\f$
+	 */
+	static SGVector<T> compute(SGMatrix<T> m, bool no_diag)
+	{
+		REQUIRE(m.num_cols > 0, "Matrix can not be empty!\n");
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/implementation/Mean.h]: Can you state that you checked the number of columns here? Also above
"Number of matrix columns (%d) must not be 0" % m.num_cols

File: src/shogun/mathematics/linalg/internal/implementation/Mean.h [hunk 9]
```diff
@@ -0,0 +1,984 @@
...
+	 * using Eigen3
+	 *
+	 * @param b the matrix-block whose colwise mean has to be computed
+	 * @param no_diag if true, diagonal entries are excluded from the mean
+	 * @return the colwise mean computed as \f$s_j=\mean_{i}b_{i,j}\f$
+	 */
+	static SGVector<T> compute(Block<SGMatrix<T> > b, bool no_diag)
+	{
+		REQUIRE(b.m_col_size > 0, "Matrix can not be empty!\n");
+                SGVector<T> result(b.m_col_size);
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/implementation/Mean.h]: whitespace error

File: src/shogun/mathematics/linalg/internal/implementation/Mean.h [hunk 10]
```diff
@@ -0,0 +1,984 @@
...
+                                for (index_t i = 0; i < mat.num_rows; ++i)
+                                        r[i] /= (mat.num_rows - 1);
+                                for (index_t i = mat.num_rows; i < mat.num_cols; ++i)
+                                        r[i] /= mat.num_rows;
+                        }
+                        else
+                        {
+                                for (index_t i = 0; i < mat.num_cols; ++i)
+                                        r[i] /= (mat.num_rows - 1);
+                                //scale<Backend::EIGEN3, Matrix>::compute(result, result, 1 / (mat.num_rows - 1));
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/implementation/Mean.h]: can you remove such outcommented code?

File: src/shogun/mathematics/linalg/internal/implementation/Mean.h [hunk 11]
```diff
@@ -0,0 +1,984 @@
...
+	 * Method that computes the row wise mean of SGMatrix using Eigen3
+	 *
+	 * @param m the matrix whose rowwise mean has to be computed
+	 * @param no_diag if true, diagonal entries are excluded from the mean
+	 * @return the rowwise mean computed as \f$s_i=\mean_{j}m_{i,j}\f$
+	 */
+	static SGVector<T> compute(SGMatrix<T> m, bool no_diag)
+	{
+		REQUIRE(m.num_rows > 0, "Matrix can not be empty!\n");
+                SGVector<T> result(m.num_rows);
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/implementation/Mean.h]: whitespace

File: src/shogun/mathematics/linalg/internal/implementation/Mean.h [hunk 12]
```diff
@@ -0,0 +1,984 @@
...
+	 * using Eigen3
+	 *
+	 * @param b the matrix-block whose rowwise mean has to be computed
+	 * @param no_diag if true, diagonal entries are excluded from the mean
+	 * @return the rowwise mean computed as \f$s_i=\mean_{j}m_{i,j}\f$
+	 */
+	static SGVector<T> compute(Block<SGMatrix<T> > b, bool no_diag)
+	{
+		REQUIRE(b.m_row_size > 0, "Matrix can not be empty!\n");
+                SGVector<T> result(b.m_row_size);
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/implementation/Mean.h]: whitespace

File: src/shogun/mathematics/linalg/internal/implementation/Mean.h [hunk 13]
```diff
@@ -0,0 +1,984 @@
...
+		return result;
+	}
+
+	/**
+	 * Method that computes the row wise mean of SGMatrix blocks
+	 * using Eigen3
+	 *
+	 * @param b the matrix-block whose rowwise mean has to be computed
+	 * @param no_diag if true, diagonal entries are excluded from the mean
+	 * @return the rowwise mean computed as \f$s_i=\mean_{j}m_{i,j}\f$
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/implementation/Mean.h]: latex. Make sure to search for it in the file :)

File: src/shogun/mathematics/linalg/internal/implementation/Mean.h [hunk 14]
```diff
@@ -0,0 +1,984 @@
...
+		std::string kernel_name = "mean_" + ocl::get_type_string<T>();
+		if (no_diag) kernel_name.append("_no_diag");
+
+		if (ocl::kernel_exists(kernel_name))
+			return ocl::get_kernel(kernel_name);
+
+		std::string source = ocl::generate_kernel_preamble<T>(kernel_name);
+		if (no_diag) source.append("#define NO_DIAG\n");
+
+		source.append(
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/implementation/Mean.h]: do we really need to totally re-implement this?
I guess so.

@lambday  ?

File: src/shogun/mathematics/linalg/internal/implementation/Mean.h [hunk 15]
```diff
@@ -0,0 +1,984 @@
...
+	/**
+	 * Method that computes the mean of symmetric CGPUMatrix using ViennaCL
+	 *
+	 * @param m the matrix whose mean has to be computed
+	 * @param no_diag if true, diagonal entries are excluded from the mean
+	 * @return the mean computed as \f$\mean_{i,j}m_{i,j}\f$
+	 */
+	static T compute(CGPUMatrix<T> mat, bool no_diag)
+	{
+		return mean<Backend::VIENNACL, CGPUMatrix<T> >::compute(mat, no_diag);
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/implementation/Mean.h]: no exploitation of symmetry here?

File: src/shogun/mathematics/linalg/internal/implementation/Mean.h [hunk 16]
```diff
@@ -0,0 +1,984 @@
...
+	 * @param no_diag if true, diagonal entries are excluded from the mean
+	 * @param result Pre-allocated vector for the result of the computation
+	 */
+	static void compute(CGPUMatrix<T> mat, CGPUVector<T> result, bool no_diag)
+	{
+                CGPUVector<T> sum_result(mat.num_cols);
+                sum_result = colwise_sum<Backend::VIENNACL, CGPUMatrix<T> >::compute(mat, no_diag);
+                if (no_diag) 
+                {
+                        REQUIRE(mat.num_rows != mat.num_cols, "Matrix is not square!\n");
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/implementation/Mean.h]: Please in error messages always give full information.
Matrix has size %dx%d, which is not square
See the dev guidelines on this here:
https://github.com/shogun-toolbox/shogun/wiki/Assertions

File: src/shogun/mathematics/linalg/internal/implementation/VectorMean.h [hunk 17]
```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright (c) The Shogun Machine Learning Toolbox
+ * Written (w) 2014 Soumyajit De
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/internal/implementation/VectorMean.h]: name :)

**Comment by karlnapf** [comment]
This is great stuff. Thanks a lot for this patch.

A few things:
 * can you please split this into multiple patches with smaller number of lines of code. There are lots of operations and this should be easy to do. At least two different ones for the vector and matrix
 * This needs unit tests before it can be merged. Have a look in `tests/unit`. You can run them as I stated in my first comment. You need to aim for >90% coverage in such fundamental things, especially with the viennacl code
 * This is why I suggest you send the methods one-by-one (one method, with unit test), so it is easier to review them and make sure no corner cases slip through.

**Comment by OXPHOS** [comment]
Thanks @karlnapf for the review! 

Thought it would be easy in the first place as it's mostly codes replacing.

The ViennaCL stuff is too much for me. I just managed to find a way to return the sizes in ViennaCL and this seems not enough..

I'll separate the patches and test by units. Thanks again!

**Comment by OXPHOS** [comment]
Hi @karlnapf I ran into problems with `cmake -DENABLE_TESTING=ON`...
I did
`--mkdir build --cd build --cmake -DENABLE_TESTING=ON .. --make`
And it stopped with 

>  shogun-ori/tests/unit/base/main_unittest.cc:1:10: fatal error: 'gtest/gtest.h' file not found
>  #include  <gtest/gtest.h>

Indeed I didn't find any gtest files.
Everything seems fine if I directly `cmake ..` 

Any idea what could be happening? Thanks!

**Comment by karlnapf** [comment]
Yeah let's do the ViennaCL after the eigen part is done :)
In general, aim for small patches, as small as possible. Rather have 10 PRs with each 15 lines than one with 150

**Comment by karlnapf** [comment]
Your cmake probem is solved?

**Comment by OXPHOS** [comment]
Yes thanks. Read some docs and it seems reading is helpful!

**Comment by OXPHOS** [comment]
Update:
Like all the other C++ based softwares I tried to run, setting things up takes the most time...

```
mkdir build
cd build
cmake -DENABLE_TESTING=on -DENABLE_VIENNACL=off -DVIENNACL_WITH_OPENCL=off ..
make
make test
make unit-tests
```

I have the following failed tests:
make test - ` 52 - unit-PeriodicKernelTest (Failed)`
make unit-tests -
```
[----------] Global test environment tear-down
[==========] 694 tests from 149 test cases ran. (3454 ms total)
[  PASSED  ] 691 tests.
[  FAILED  ] 3 tests, listed below:
[  FAILED  ] PeriodicKernelTest.test_kernel_matrix
[  FAILED  ] PeriodicKernelTest.test_derivative_width
[  FAILED  ] PeriodicKernelTest.test_derivative_period
```
-------------------------------------------------------------------------------------------------------------------------------------------
For the part I implemented, I modified `VectorMean_unittest.cc` from `VectorSum_unittest.cc`.
The test passed. However, I noticed, for both  `VectorMean_unittest.cc` and `VectorSum_unittest.cc`, only Native backend were tested. I suppose Eigen3 is more important now.

```
[----------] 1 test from VectorMean
[ RUN      ] VectorMean.native_backend
[       OK ] VectorMean.native_backend (0 ms)
[----------] 1 test from VectorMean (0 ms total)

[----------] 1 test from VectorSum
[ RUN      ] VectorSum.native_backend
[       OK ] VectorSum.native_backend (0 ms)
[----------] 1 test from VectorSum (0 ms total)
```

-------------------------------------------------------------------------------------------------------------------------------------------

I checked `CMakeCache`. Eigen3 should be available:
```
//Details about finding Eigen3
FIND_PACKAGE_MESSAGE_DETAILS_Eigen3:INTERNAL=[/usr/local/include/eigen3][v3.2.8(3.1.2)]
```
So I was not able to test Eigen3 backend so far.

-------------------------------------------------------------------------------------------------------------------------------------------
For the Latex part, I haven't spent too much effort.
Did you mean I should change from `\f$\sum_i a_i\f$` to `\f$\frac{1}{N^2}\sum_{i,j}\f$`? I suppose the case is different for vector mean? And I noticed somebody mentioned \mean_xxx\ can represent bar. So this is not fixed yet.

**Comment by karlnapf** [comment]
@OXPHOS yes definitely, the unit tests native should be replaced with eigen3. In your case, just add tests for eigen3. Once you have done so, send them one-by-one in new pull requests -- travis will run all your added unit tests and we can see whether things worked. I will close this one for now  -- looking forward to the new one! :) (might give a few comments on the update you made)

Really relevant, so I ping the patches that removed things. Please when removing something, always replace the unit tests and implementations for eigen3. We dont want to cut out features here. This is not for you @OXPHOS 
#3081 #3080

**Comment by OXPHOS** [comment]
@karlnapf Thanks but don't bother looking at this one. I'll open a new pull for VectorMean-Eigen3.
